mixins/provider-mixin.js

```diff
@@ -0,0 +1,38 @@
+export const ProviderMixin = superclass => class extends superclass {
+
+	constructor() {
+		super();
+		this._providers = new Map();
+		this.addEventListener('d2l-request-provider', this._handleRequest);
+	}
+
+	_handleRequest(e) {
+		if (this._providers.has(e.detail.key)) {
+			e.detail.provider = this._providers.get(e.detail.key);
```

Comment:
Is there any chance that a dependency is provided async?  For example, consumer asks for a view-dependency that is dynamically loaded.  Perhaps `e.detail.provider` should be a promise for the dependency.  Then `requestProvider` (below) can either be async and wait-for/return the dependency, or it could return the promise to the caller.

Reply:
I thought about this as well but then didn't mention it. It being synchronous has some nice behaviours... requesters can request things in their `render()` methods for example, or anywhere really and be guaranteed that things will exist when they render. If things become `async`, that complicates things a lot.

IMO, we leave this synchronous and then if we run into a case where we need dependency injection of things that are ayns, we introduce an `AsyncProviderMixin` or add a `requestDependencyAsync()` or something.

Reply:
Yeah, I think the `Promise` is a good call - the event handling is synchronous, but since we're returning a function anyway, might as well just return a Promise instead.